cmd/collector/app/sanitizer_v2/empty_service_name_sanitizer.go

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package sanitizer_v2
+
+import (
+	"go.opentelemetry.io/collector/pdata/ptrace"
+)
+
+// Constants for the replacement names
+const (
+	serviceNameReplacement = "empty-service-name"
+	nullProcessServiceName = "null-process-and-service-name"
+)
+
+// NewEmptyServiceNameSanitizer returns a function that replaces empty service names
+// with a predefined string.
+func NewEmptyServiceNameSanitizer() SanitizeSpan {
+	return sanitizeEmptyServiceName
+}
+
+// sanitizeEmptyServiceName sanitizes the service names in the span attributes.
+func sanitizeEmptyServiceName(span ptrace.Span) ptrace.Span {
+	attributes := span.Attributes()
+	serviceNameAttr, ok := attributes.Get("service.name")
```

Comment:
service name is stored in the Resource attributes, not in the span.

Reply:
@yurishkuro` span.Resource` undefined (type ptrace.Span has no field or method Resource)compilerMissingFieldOrMethod

This means we need to access the resource from the` ptrace.ResourceSpans`. I think we need to work with `ptrace.Traces `instead of individual `ptrace.Span` and access the ResourceSpans which contains the Resource and associated ScopeSpans. Then we need to iterate through each ScopeSpans and their Spans. Am I right?

Reply:
> I think we need to work with ptrace.Traces 

@james-ryans this is an interesting situation. In v1 pipeline all non-span data (like batch.Process) was first denormalized into each span so that the unit of work in the pipeline was a single span. In v2 storage API we want to support batching for better efficiency, but it would also mean that unless storage supports storing normalized data (like storing resource / scope records separately, which is unlikely in k/v stores world), each storage would need to do its own denormalization. It also creates challenges to shared capabilities like sanitizers here, since they need to operate on data uniformly before the storage is called.

We can provide a shared lib to do the flattening / denormalization of resources[]/scopes[]/spans[] into just spans[] (essentially otlp2jaeger does that today already), but I don't think it's a good solution considering that OTLP spans cannot distinguish between resource/scope/span attributes without some magic prefixes / naming. In v1 model we had a clear slot Span.Process to maintain that distinction. I am thinking that in v2 storage implementations their internal data model should also be `Span{ Resource=..., Scope=..., rest of span }`, to avoid unnecessary flattening.

Coming back to sanitizers, I think `Sanitize(pdata.Traces)` is a good starting point signature. Some sanitizers would only look at Resources in that input, while others might look at everything. The alternative would be to have distinct sanitizers for Resource, Scope, and Span, but I don't think we need this complexity at this point. 

@varshith257 does this answer your question?

Reply:
>does this answer your question?

Yes. Thanks for it